### PR TITLE
Integrate ACL-FP16 into oneDNN

### PR DESCRIPTION
**Pull request by lilh9598**
# Description
This PR is for integrating ACL-FP16 into oneDNN.  Feel free to let me know if there is any code or document that I should supplement. 

And based on these two facts as follows, I think it makes sense to integrate ACL-FP16 into oneDNN. 
- For hardware: Most Aarch64 CPUs support FP16 Instruction Set Architecture.
- For software: Most ACL operators could work with data type FP16.

# Checklist

## General

- [x] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [x] Have you formatted the code using clang-format?

### CODE REVIEW

**Comment by igorsafo** [comment]
Hi @lilh9598 , thanks for the PR! Overall the changes look good to me. Here are some comments:
- Please apply clang-format to your changes.
- New example fails on x64 CPU due to unimplemented functionality. Please add handling for this issue similar to bf16 example: https://github.com/oneapi-src/oneDNN/blob/62c7bc5f5c73133f746868663c29f45bb0492b64/examples/cnn_training_bf16.cpp#L110-L121
- I think example is not enough to validate the changes. So we need to make sure benchdnn works with fp16 for Aarch64. Gtests is another part.


@nSircombe @cfRod @jondea  Could you please take a look at the changes?

@kawakami-k just for your information.

**Comment by lilh9598** [comment]
> Hi @lilh9598 , thanks for the PR! Overall the changes look good to me. Here are some comments:
> 
> * Please apply clang-format to your changes.
> * New example fails on x64 CPU due to unimplemented functionality. Please add handling for this issue similar to bf16 example: https://github.com/oneapi-src/oneDNN/blob/62c7bc5f5c73133f746868663c29f45bb0492b64/examples/cnn_training_bf16.cpp#L110-L121
> * I think example is not enough to validate the changes. So we need to make sure benchdnn works with fp16 for Aarch64. Gtests is another part.
> 
> @nSircombe @cfRod @jondea Could you please take a look at the changes?
> 
> @kawakami-k just for your information.

@igorsafo Thanks! As you suggested, I'm going to code gtests and benchmarks next.

**Comment by nSircombe** [comment]
Hi @lilh9598, thanks for submitting this PR!
@igorsafo - thanks for the heads-up we'll take a look.

File: src/cpu/aarch64/acl_binary.hpp [hunk 1]
```diff
@@ -129,7 +129,8 @@ struct acl_binary_t : public primitive_t {
 
             // Only support f32 and s32 for now
```
Comment (nSircombe) [inline src/cpu/aarch64/acl_binary.hpp:130]: This comment will need to be updated now.

File: src/cpu/aarch64/acl_eltwise.hpp [hunk 2]
```diff
@@ -74,11 +74,17 @@ struct acl_eltwise_fwd_t : public primitive_t {
             using namespace data_type;
             const memory_desc_wrapper data_d(data_md());
 
-            bool ok = is_fwd() && one_of(data_d.data_type(), f32, s32, s8)
+            bool ok = is_fwd() && one_of(data_d.data_type(), f32, f16, s32, s8)
                     && !has_zero_dim_memory() && attr()->has_default_values()
                     && data_d.is_dense();
             if (!ok) return status::unimplemented;
 
+            // Workaround for the nan-value caused by tanh of ACL-FP16 int gtest(test_eltwise).
```
Comment (nSircombe) [inline src/cpu/aarch64/acl_eltwise.hpp:82]: It would be worth raising this as an issue against ACL I think

File: src/cpu/platform.cpp [hunk 3]
```diff
@@ -119,6 +119,8 @@ bool has_data_type_support(data_type_t data_type) {
         case data_type::f16:
 #if DNNL_X64
             return x64::mayiuse(x64::avx512_core_fp16);
+#elif DNNL_AARCH64_USE_ACL
```
Comment (nSircombe) [inline src/cpu/platform.cpp:122]: It's worth noting that fp16 support in ACL will be dependent on how the ACL lib was built.

**Review by nSircombe — COMMENTED** [review]
Hi @lilh9598,

Looks really promising! I made a few comments as I went through but mostly nit-picking. In addition, I think all the copyright headers on these files will need to be updated to make sure the attribution is correct.

I was able to try out the PR on an Arm machine with fp16 support (a Neoverse-N1) on some BenchDNN batches with `cfg` set to `f16`, the `ONEDNN_VERBOSE` output, combined with a dump of the Compute Library kernel name, clearly shows it calling into ACL, and dispatching to an fp16 Neon kernel (`a64_hybrid_fp16_mla_6x32` for example).

There are a few things worth noting though:

- The availability of fp16 kernels in ACL is dependent on the build flags employed, noting that arm-v8a  does not have fp16 support, so a generic build may not provide fp16 support. For my testing I build ACL with `multi_isa` support, which should work on v8a and above but provide fp16, sve and bf16 kernels where hardware permits (build line being `scons -j32 Werror=0 debug=0 neon=1 embed_kernels=0 os=linux arch=armv8a multi_isa=1 build=native`).

- I tried the same build on an older, A72, platform without fp16 support and ACL returned `unimplemented` during the config. phase, leaving oneDNN to pick up the ref. impl. instead - which seams like the right response - @igorsafo, would that be expected behaviour in the case where HW support is lacking?

- Only a subset of the DroneCI stages use an ACL build with `multi_isa` support, as it requires gcc 10.2 or later. Also, I'm not sure if the DroneCI runners will have fp16 support, so fp16 tests in the CI may fall though to the ref. impl. and the ACL backend may not be exercised.

- There are limitations on the use of the weights tensor in ACL 22.05 and older. These versions do not allow the weights tensor to be re-written after primitive creation - this means that primitive caching cannot be used (there are fixes in TensorFlow to prevent such caching of ACL based oneDNN primitives at present). The 22.08 release introduces a new memory format with allows the GEMM kernels to ingest the weights directly from the wei tensor, skipping the intermediate interleaving step in ACL and allowing wei to be re-written at will. We're working to integrate this support into TensorFlow at the moment, which will allow the removal of the kind of workarounds we currently have in place to prevent it from caching ACL primitives. At present, support for these new memory formats hasn't been exposed in ACL for fp16, so some work would be needed to enable this.
None of this has any immediate impact on this oneDNN PR, but it's worth bearing in mind if you're looking to integrate support into a framework like TensorFlow.

Finally, it would probably be good to get another set of eyes from the Arm team to take a look - I'll give someone a nudge!

File: src/cpu/aarch64/acl_eltwise.hpp (hunk 2 above)
Comment (lilh9598) [reply src/cpu/aarch64/acl_eltwise.hpp:82]: You are right, I will code a program that can reproduce this bug for ACL later.

File: src/cpu/platform.cpp (hunk 3 above)
Comment (lilh9598) [reply src/cpu/platform.cpp:122]: Yes, as you said, we need to enable multi_isa options.

**Comment by lilh9598** [comment]
> * I tried the same build on an older, A72, platform without fp16 support and ACL returned `unimplemented` during the config. phase, leaving oneDNN to pick up the ref. impl. instead - which seams like the right response - @igorsafo, would that be expected behaviour in the case where HW support is lacking?

Hi @nSircombe,  

Thanks for your careful review!   
  
The problem which appear on the  platform without fp16 ISA support can be solve with  'arm_compute::CPUInfo::get().has_fp16()' I think.  The change is as follows:  
`case data_type::f16: return arm_compute::CPUInfo::get().has_fp16();`

File: src/cpu/platform.cpp [hunk 4]
```diff
@@ -119,6 +120,8 @@ bool has_data_type_support(data_type_t data_type) {
...
+            return arm_compute::CPUInfo::get().has_fp16();
```
Comment (lilh9598) [inline src/cpu/platform.cpp:124]: @nSircombe

**Comment by nSircombe** [comment]
HI @lilh9598,

> The problem which appear on the platform without fp16 ISA support can be solve with
> 'arm_compute::CPUInfo::get().has_fp16()' I think. The change is as follows:
> case data_type::f16: return arm_compute::CPUInfo::get().has_fp16();

Sorry, my comment could be clearer - actually, I don't _think_ there is a problem here. At the moment this check falls to ACL at the validate phase, on a machine without fp16 support, validation will fail in ACL, and return `unimplemented`, causing oneDNN to fall back gracefully to the reference kernel. This looks good to me, but I'm not _entirely_ sure what's expected. Should this in fact generate an error in oneDNN, for example?

**Comment by lilh9598** [comment]
> 
I get your point, it is really necessary to make clear what's expected in oneDNN.

**Comment by lilh9598** [comment]
> Sorry, my comment could be clearer - actually, I don't _think_ there is a problem here. At the moment this check falls to ACL at the validate phase, on a machine without fp16 support, validation will fail in ACL, and return `unimplemented`, causing oneDNN to fall back gracefully to the reference kernel. This looks good to me, but I'm not _entirely_ sure what's expected. Should this in fact generate an error in oneDNN, for example?
> 
> ...as noted though, this might have implications for the CI, but it would always be possible to employ QEMU, as the SVE stages do, if fp16 isn't available on the DroneCI machines.

I get your point, it is really necessary to make clear what's expected in oneDNN.

File: src/cpu/aarch64/acl_inner_product.hpp [hunk 5]
```diff
@@ -37,6 +37,7 @@ struct acl_ip_obj_t {
 
 struct acl_ip_conf_t {
     bool with_bias;
+
```
Comment (cfRod) [inline src/cpu/aarch64/acl_inner_product.hpp:40]: Extra line?

**Comment by lilh9598** [comment]
Hi all @nSircombe @igorsafo @cfRod , 
I do all benchdnn tests on my machine with fp16 support and get the four failed cases.

> 108:test_benchdnn_eltwise_ci_cpu
119:test_benchdnn_gru_ci_cpu
120:test_benchdnn_lstm_ci_cpu
121:test_benchdnn_rnn_ci_cpu
  

- The failure of test_benchdnn_eltwise_ci_cpu is due to a strict precision limit, some of the test results are shown below.

> [   1][DST][0:0:0:0:1] exp_f32:0.00247569 exp:0.00247574 got:0.00292587 diff:0.000450134 rdiff:0.181818
[  27][DST][0:0:1:4:0] exp_f32:0.00247569 exp:0.00247574 got:0.00292587 diff:0.000450134 rdiff:0.181818
[  40][DST][0:1:0:3:1] exp_f32:0.0181499 exp:0.0181427 got:0.0183868 diff:0.000244141 rdiff:0.0134567
[  85][DST][0:2:1:3:1] exp_f32:0.421934 exp:0.421875 got:0.422363 diff:0.000488281 rdiff:0.00115741
[  93][DST][0:3:0:1:0] exp_f32:0.696382 exp:0.696289 got:0.697266 diff:0.000976562 rdiff:0.00140252
[ 111][DST][0:3:1:2:0] exp_f32:0.355105 exp:0.355225 got:0.354736 diff:0.000488281 rdiff:0.00137457
[ 119][DST][0:3:1:4:2] exp_f32:0.703562 exp:0.703613 got:0.702637 diff:0.000976562 rdiff:0.00138793
[ 131][DST][0:4:0:3:2] exp_f32:0.126928 exp:0.126953 got:0.127319 diff:0.000366211 rdiff:0.00288462
[ 170][DST][0:5:1:1:2] exp_f32:0.00671535 exp:0.00671387 got:0.00681305 diff:9.91821e-05 rdiff:0.0147727

- The failure of the other three cases is due to the absence of reference kernel(rnn/lstm/gru) with fp16, and I find out that there is no implementation with fp16 in the cpu_rnn_list.
https://github.com/oneapi-src/oneDNN/blob/12d0fd478a906e47bd4b052071a29716f54b55b7/src/cpu/cpu_rnn_list.cpp#L32-L37

Some of the test results in test_benchdnn_rnn_ci_cpu are shown below:
> run: --rnn --cfg=f16 --activation=TANH l1t1mb1sic16slc16dhc16dic16n"uniform"
@@@ error [int create_primitive(benchdnn_dnnl_wrapper_t<dnnl_primitive*>&, dnnl_engine_t, const func_t&, const prb_t*, res_t*, dir_t, const_dnnl_primitive_desc_t, bool) [with func_t = dnnl_status_t(init_pd_args_t<rnn::prb_t>&); prb_t = rnn::prb_t; dnnl_engine_t = dnnl_engine*; const_dnnl_primitive_desc_t = const dnnl_primitive_desc*]:543]: 'check_dnnl_status(status, prb, res)' -> 1
@@@ error [int init_prim(benchdnn_dnnl_wrapper_t<dnnl_primitive*>&, const func_t&, const prb_t*, res_t*, dir_t, const_dnnl_primitive_desc_t, bool) [with func_t = dnnl_status_t(init_pd_args_t<rnn::prb_t>&); prb_t = const rnn::prb_t; const_dnnl_primitive_desc_t = const dnnl_primitive_desc*]:592]: 'create_primitive(primw, engine, init_pd_func, prb, res, dir, hint, is_service_prim)' -> 1
@@@ error [int rnn::doit(const rnn::prb_t&, res_t*):857]: 'init_prim(prb.ctx_init, prim, init_pd, &prb, res, FLAG_FWD, nullptr, is_service_prim)' -> 1
168:UNIMPLEMENTED __REPRO: --rnn --cfg=f16 --activation=TANH l1t1mb1sic16slc16dhc16dic16n"uniform"

Could you give me some advice about how to handle the mentioned failures？

File: src/cpu/platform.cpp [hunk 6]
```diff
@@ -36,6 +36,7 @@
 #include "cpu/aarch64/cpu_isa_traits.hpp"
 #if DNNL_AARCH64_USE_ACL
 // For setting the number of threads for ACL
+#include "arm_compute/core/CPP/CPPTypes.h"
```
Comment (cfRod) [inline src/cpu/platform.cpp:39]: Could you update the comment on line 38?

Comment (lilh9598) [reply src/cpu/platform.cpp:39]: okk

File: src/cpu/aarch64/acl_pooling.hpp [hunk 7]
```diff
@@ -74,8 +74,8 @@ struct acl_pooling_fwd_t : public primitive_t {
         status_t init(engine_t *engine) {
             bool ok = set_default_params() == status::success
                     && is_fwd() // ACL supports forward propagation only
-                    && utils::everyone_is(data_type::f32, src_md()->data_type,
-                            dst_md()->data_type)
```
Comment (cfRod) [inline src/cpu/aarch64/acl_pooling.hpp:78]: The check for dst_md is missing?

Comment (lilh9598) [reply src/cpu/aarch64/acl_pooling.hpp:78]: yeah, I missed the check

File: src/cpu/aarch64/acl_eltwise.hpp (hunk 2 above)
Comment (lilh9598) [reply src/cpu/aarch64/acl_eltwise.hpp:82]: https://github.com/ARM-software/ComputeLibrary/issues/998

**Comment by mgouicem** [comment]
>  This looks good to me, but I'm not entirely sure what's expected. Should this in fact generate an error in oneDNN, for example?

This seems good to me, and this is why we have a reference path at all :-)

> The failure of test_benchdnn_eltwise_ci_cpu is due to a strict precision limit, some of the test results are shown below.

We are usually ok to relax a little the error bounds in eltwise checks. However, some of the errors here seem large when looking at relative difference. Could you specify which algorithms show these inaccuracies?

> The failure of the other three cases is due to the absence of reference kernel(rnn/lstm/gru) with fp16, and I find out that there is no implementation with fp16 in the cpu_rnn_list.

We typically filter those out in benchdnn [here](https://github.com/oneapi-src/oneDNN/blob/f03046f2331a586905b2d41ede028af52ffe9590/tests/benchdnn/rnn/rnn.cpp#L678).
You can update the rule [here](https://github.com/oneapi-src/oneDNN/blob/f03046f2331a586905b2d41ede028af52ffe9590/tests/benchdnn/rnn/rnn.cpp#L695-L703) to not run rnn problems in f16 on ARM.

File: src/cpu/aarch64/acl_indirect_gemm_convolution.cpp [hunk 8]
```diff
@@ -21,8 +21,10 @@ namespace impl {
 namespace cpu {
 namespace aarch64 {
 
-status_t acl_indirect_gemm_convolution_fwd_t::execute_forward(
-        const exec_ctx_t &ctx) const {
+template <data_type_t src_type, data_type_t wei_type, data_type_t dst_type,
```
Comment (jondea) [inline src/cpu/aarch64/acl_indirect_gemm_convolution.cpp:24]: Hi @lilh9598, thanks for your contribution!

I know this is an existing pattern for `acl_gemm_convolution_fwd_t`, but we don't need to template this class, we can handle the different types at runtime like in ACL binary and eltwise. Removing the templating should reduce build time and build size.

The same applies to winograd and inner product.

**Comment by lilh9598** [comment]
> We are usually ok to relax a little the error bounds in eltwise checks. However, some of the errors here seem large when looking at relative difference. Could you specify which algorithms show these inaccuracies?
Hi @mgouicem ,  
Thanks for you comment! The three algorithms 'logistic/soft_relu/elu' show these inaccuracies, and all the test log in 
[Uploading LastTest.log…]()

**Comment by lilh9598** [comment]
https://github.com/ARM-software/ComputeLibrary/blob/aabef6c0584f06f4c0f4b61fb787d80374240619/src/core/NEON/NEMath.inl#L511-L518

https://github.com/ARM-software/ComputeLibrary/blob/aabef6c0584f06f4c0f4b61fb787d80374240619/src/core/NEON/NEMath.inl#L145-L167

File: src/cpu/aarch64/acl_indirect_gemm_convolution.cpp (hunk 8 above)
Comment (lilh9598) [reply src/cpu/aarch64/acl_indirect_gemm_convolution.cpp]: Thanks for your review! I made a fix.

**Comment by mgouicem** [comment]
> Hi @mgouicem , thanks for your comment! The three algorithms 'logistic/soft_relu/elu' show these inaccuracies, and all the test log in attachment. [LastTest.log.zip](https://github.com/oneapi-src/oneDNN/files/9618412/LastTest.log.zip).

Thanks for the details.

> ACL calculates exp by [this approximate algorithm](https://github.com/ARM-software/ComputeLibrary/blob/aabef6c0584f06f4c0f4b61fb787d80374240619/src/core/NEON/NEMath.inl#L145-L167), maybe that's why there are errors with fp16 on 'logistic/soft_relu/elu' which contains exp. https://github.com/ARM-software/ComputeLibrary/blob/aabef6c0584f06f4c0f4b61fb787d80374240619/src/core/NEON/NEMath.inl#L511-L518

From a quick look, the exp implementation seems ok. This is very similar to what oneDNN implements for x64 [here](https://github.com/oneapi-src/oneDNN/blob/master/src/cpu/x64/injectors/jit_uni_eltwise_injector.cpp#L260).
I also tried to exhaustively check it and it seemed not too bad (worst case relative error was ~2^-18).

I guess the error might be with how exp is used in logistic/soft_relu/elu. In particular, they might require dedicated implementation of log1p or expm1 to avoid cancellations around 0.

**Comment by lilh9598** [comment]
> I guess the error might be with how exp is used in logistic/soft_relu/elu. In particular, they might require dedicated implementation of log1p or expm1 to avoid cancellations around 0.

The implementation of logistic/soft_relu/elu [here](https://github.com/ARM-software/ComputeLibrary/blob/aabef6c0584f06f4c0f4b61fb787d80374240619/src/cpu/kernels/activation/generic/neon/impl.h#L105-L107) looks ok,  these inaccuracies are due to fp16 I think, but similar way works fine in fp32

File: examples/cnn_inference_f16.cpp [hunk 9]
```diff
@@ -0,0 +1,806 @@
+/*******************************************************************************
```
Comment (densamoilov) [inline examples/cnn_inference_f16.cpp:1]: I think adding a new example for f16 data type is redundant as the only difference between `cnn_inference_f16.cpp` and `cnn_inference_f32.cpp` would be using different data types `f16` vs `f32`. It looks like it's not worth having the whole new example.

Comment (lilh9598) [reply examples/cnn_inference_f16.cpp:1]: Thanks for the review, I try to code a more reasonable example about f16.

Comment (lilh9598) [reply examples/cnn_inference_f16.cpp]: I deleted this file, I think that we need a new PR to rework the cnn_inference_example for f16/bf16/f32 data type.

**Review by densamoilov — APPROVED** [review]

File: examples/cnn_inference_f16.cpp (hunk 9 above)
Comment (densamoilov) [reply examples/cnn_inference_f16.cpp]: Ok, just keep in mind that we usually provide examples to demonstrate use of our API, though it's not a strict requirement.
A an example, we provide examples f32 and int8, the difference is that for int8 we demonstrate use of our output scale API, that is necessary for int8 data types.

**Comment by densamoilov** [comment]
@lilh9598, is it ready to be merged?

**Comment by lilh9598** [comment]
> @lilh9598, is it ready to be merged?

@densamoilov ，no yet，I'm thinking about how to deal with the inaccuracies caused by logistic/soft_relu/elu of ACL-FP16

> @lilh9598, is it ready to be merged?

@densamoilov , no yet

**Comment by lilh9598** [comment]
I add ACL prelu/bn/softmax/matmul for data type fp16 compared to before, @jondea @cfRod  could you take a look again?

**Comment by jondea** [comment]
I've created an issue on the ACL repo for this. In the meantime, is it worth returning `unimplemented` for any of the failing activations and coming back to this later when it is fixed?

**Comment by lilh9598** [comment]
> I've created an issue on the ACL repo for this. In the meantime, is it worth returning `unimplemented` for any of the failing activations and coming back to this later when it is fixed?

Hi @jondea ,
The error(nan-value) of tanh for fp16 was fixed [here](https://github.com/ARM-software/ComputeLibrary/issues/998), the high relative errors are caused by logistic/soft_relu/elu of ACL for fp16.  When the errors are fixed, the ref primitive should be replaced with the primitive of ACL I think.

**Comment by jondea** [comment]
The fix for the overflows and the high relative error in FP16 tanh will be in the ACL 22.11 release. However, currently oneDNN enforces 22.05 and it's likely that we won't manage to get ACL 22.11 enforced for oneDNN 3.0 (because the releases are close) so I think returning unimplemented is needed. 

oneDNN 3.1 should enforce ACL 22.11, so we can remove it then.

**Comment by lilh9598** [comment]
> The fix for the overflows and the high relative error in FP16 tanh will be in the ACL 22.11 release. However, currently oneDNN enforces 22.05 and it's likely that we won't manage to get ACL 22.11 enforced for oneDNN 3.0 (because the releases are close) so I think returning unimplemented is needed.
> 
> oneDNN 3.1 should enforce ACL 22.11, so we can remove it then.

Reasonable solution.

> The fix for the overflows and the high relative error in FP16 tanh will be in the ACL 22.11 release. However, currently oneDNN enforces 22.05 and it's likely that we won't manage to get ACL 22.11 enforced for oneDNN 3.0 (because the releases are close) so I think returning unimplemented is needed.
> 
> oneDNN 3.1 should enforce ACL 22.11, so we can remove it then.

@densamoilov，I think it is ready to be merged if there are no other comments.

**Review by mgouicem — APPROVED** [review]
LGTM thanks for the PR!

**Comment by dzarukin** [comment]
Hi @lilh9598, could you please rebase this PR on top of latest master? Thank you.

**Comment by lilh9598** [comment]
> Hi @lilh9598, could you please rebase this PR on top of latest master? Thank you.

@dzarukin Done, thanks for reminding me.

**Comment by dzarukin** [comment]
Thank you for the contribution!

**Pull request merged**